### ISSUE TITLE
Test test_mainwindow.py hangs

### ISSUE THREAD

**Issue by juliangilbey** [spyder-ide/spyder#21189]
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ ] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description

This is a somewhat bizarre one; I'm asking just in case you have any idea.  As you may know, we run our CI tests on Debian using lxc.  Because the `test_mainwindow.py` tests are sometimes flaky, I run them on their own (so I run the main tests with `--deselect=spyder/app/tests/test_mainwindow.py` and then test `test_mainwindow.py` separately, once with `--run-slow` and once without).

This worked fine with 5.4.2, but now that I'm trying to package 5.4.4, I'm finding that `test_mainwindow.py` with `--run-slow` typically gets to 99% and then just hangs.  Running it a second time (using the `pytest_log.txt` to skip previously passed tests) then goes fine - every individual test passes.  But it seems that something changed between 5.4.2 and 5.4.4 that means that something is not finishing properly.  (5.4.3 was released during the Debian freeze, so I didn't try packaging that.)  I tried excluding the test where the process hung, but that didn't help - it just hung on a different test instead.

Do you have any idea what might be causing this?  I am trying to work around it using `timeout` (though frustratingly that doesn't seem to be working inside the `lxc` container at the moment for some reason I don't understand), but solving the root cause would obviously be preferable.

## Versions

* Spyder version: 5.4.4
* Python version: 3.11.4
* Qt version: 5.15.9
* PyQt version: 5.15.9
* Operating System name/version: Debian testing

(Edit: BTW, `test_mainwindow.py` without `--run-slow` of course does nothing!)

**Comment by ccordoba12** [comment]
Our `5.x` tests have been running fine for several months and I don't see any problems when looking at the Linux Github actions logs:

https://github.com/spyder-ide/spyder/actions/runs/5661429629/job/15339300204

But I did some changes to the way we run our tests on PR https://github.com/spyder-ide/spyder/pull/20425 (included in 5.4.3), which removed the `slow` marker. Instead, we break the test suite as described here:

https://github.com/spyder-ide/spyder/pull/20425/files#diff-a31c7ed5d35f5ed8233994868c54d625b18e6bacb6794344c4531e62bd9dde59R70

But that means that in your case you don't need to

> then test test_mainwindow.py separately, once with --run-slow and once without

If the main window tests are passing without `--run-slow`, then they are passing as expected.

**Comment by juliangilbey** [comment]
(Not quite; the `--run-slow` option is still needed to run the main window tests; they're all marked as "slow", but in `conftest.py` rather than individually in `test_mainwindow.py`.)

I had guessed it was some sort of local problem, as I'd seen the actions were passing on GitHub.  I just thought that maybe you would have a clue what might be causing it.  But if not, feel free to close this.

**Comment by ccordoba12** [comment]
> (Not quite; the --run-slow option is still needed to run the main window tests; they're all marked as "slow", but in conftest.py rather than individually in test_mainwindow.py.)

I meant that you can run them separately with

    pytest spyder/app/tests/test_mainwindow.py -vv

In that case, `--run-slow` is not necessary (it does nothing because there are no tests marked as slow, as I said).

How are you running them?

**Comment by juliangilbey** [comment]
Oh.  We're running the following command:
```
    export CI=true
    export QTCONSOLE_TESTING=true

    for slow in "" "--run-slow"
    do
        debian/tests/run_pytest \
          xvfb-run -a -s "-screen 0 1024x768x24 +extension GLX" \
          $timeout_cmd $py runtests.py $EXCLUDE $slow spyder/app/tests/test_mainwindow.py
    done
```
where `debian/tests/run_pytest` is a wrapper script that runs the test up to five times, using the `pytest_log.txt` mechanism to record previously successful tests.
Interestingly though, when the test is run without `--run-slow`, nothing is tested, hence my comment above:
```
*** STARTING RUN 1 at 10:37:55: xvfb-run -a -s -screen 0 1024x768x24 +extension GLX timeout -k 5 1200 python3.11 runtests.py --deselect=spyder/plugins/editor/widgets/tests/test_hints_and_calltips.py::test_get_hints --deselect=spyder/app/tests/test_mainwindow.py::test_run_cython_code --deselect=spyder/plugins/findinfiles/widgets/tests/test_widgets.py::test_no_empty_file_items --deselect=spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_tk_backend spyder/app/tests/test_mainwindow.py
Pytest Arguments: ['-vv', '-rw', '--durations=10', '-W ignore::UserWarning', '--timeout=120', '--deselect=spyder/plugins/editor/widgets/tests/test_hints_and_calltips.py::test_get_hints', '--deselect=spyder/app/tests/test_mainwindow.py::test_run_cython_code', '--deselect=spyder/plugins/findinfiles/widgets/tests/test_widgets.py::test_no_empty_file_items', '--deselect=spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py::test_tk_backend', 'spyder/app/tests/test_mainwindow.py']
============================= test session starts ==============================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0 -- /usr/bin/python3.11
cachedir: .pytest_cache
PyQt5 5.15.9 -- Qt runtime 5.15.8 -- Qt compiled 5.15.8
rootdir: /tmp/autopkgtest-lxc.khxxpya5/downtmp/build.B2K/src
configfile: pytest.ini
plugins: cov-4.1.0, lazy-fixture-0.6.3, order-1.0.1, qt-4.2.0+repack, timeout-2.1.0, mock-3.11.1, flaky-3.7.0
timeout: 120.0s
timeout method: signal
timeout func_only: False
collecting ... collected 120 items / 1 deselected / 119 selected

spyder/app/tests/test_mainwindow.py::test_single_instance_and_edit_magic SKIPPED [  0%]
spyder/app/tests/test_mainwindow.py::test_default_plugin_actions SKIPPED [  1%]
spyder/app/tests/test_mainwindow.py::test_dedicated_consoles SKIPPED     [  2%]
spyder/app/tests/test_mainwindow.py::test_change_types_in_varexp SKIPPED [  3%]
spyder/app/tests/test_mainwindow.py::test_preferences_change_interpreter SKIPPED
 [  4%]
spyder/app/tests/test_mainwindow.py::test_no_update_outline SKIPPED      [  5%]
spyder/app/tests/test_mainwindow.py::test_leaks SKIPPED (Need --run-slow
option to run)                                                           [  5%]
spyder/app/tests/test_mainwindow.py::test_lock_action SKIPPED (Need
--run-slow option to run)                                                [  6%]
[...]
================ 119 skipped, 1 deselected, 8 warnings in 0.63s ================
*** END OF RUN 1: ALL TESTS RUN HAVE NOW PASSED/XFAILED ***
```

**Comment by juliangilbey** [comment]
Ah, I think it's because we've set `CI=true`, so `conftest.py` marks all of the tests as slow!

**Comment by ccordoba12** [comment]
Yep, that's what I was thinking too.

**Comment by juliangilbey** [comment]
Well, I did a little more investigation and the results are bizarre, to say the least. The test fails on the first attempt at `test_clickable_ipython_tracebacks` at the line
```python
    assert QApplication.overrideCursor().shape() == Qt.PointingHandCursor
```
so it attempted to try again (because the test is marked flaky).  Unfortunately, it then hangs right at the start, specifically at the line:
```python
    qtbot.waitUntil(lambda: shell._prompt_html is not None,
                    timeout=SHELL_TIMEOUT)
```
Running just this one test on its own works fine.  Removing this test makes the whole test file succeed without hanging.

But adding an extra test at the end of the file also works:
```python
@flaky(max_runs=10)                                                           
def test_null_test(main_window, qtbot, tmp_path):                             
    pass                                                                  
```
(The `@flaky` and fixtures are there to make the signature of the test look similar to that of the clickable test.)   In this case, the first run through `test_clickable_ipython_tracebacks` still fails in exactly the same way, the second attempt (controlled by flaky) works without a problem.

So there's some really strange interaction going on somewhere.  No expectation that you'll be able to solve it, and I don't think it's worth more effort, but I just wanted to say what I had discovered.
